1-theory/2-different-paradigms.adoc

```diff
@@ -0,0 +1,217 @@
+
+*Client:* Yes, please let me know who is there!
+
+*REST:* Doctors Smith and Jones, here are links to their profiles.
+
+*Client:* Ok, Doctor Jones looks like my sort of Doctor, I would like to see them, let’s make that appointment.
+
+*REST:* Appointment created, here’s a link to the appointment details.
+
+REST provided all of the relevant information with the response, and the client was able to pick through the options to resolve the situation.
+
+None of this is magic, no client is going to know exactly what to do without being trained, but the client of a REST API can be told to follow the `"alternative_doctors": "https://api.example.com/available_doctors?available_at=2017-01-01 03:00:00 GMT"`` link. That is far less of a burden on the client than expecting it to check the calendar itself, seek for availability, etc.
+
+This centralization of state into the server has benefits for systems with multiple different clients who offer similar workflows. Instead of distributing all the logic, checking data fields, showing lists of "Actions", etc. around various clients — who might come to different conclusions — REST keeps it all in one place.
+
+This book wil get more in depth on hypermedia controls later. There are a few other important things to understand about REST APIs first:
+
+- REST must be stateless: not persisting sessions between requests
+- Responses should declare cacheablility: helps your API scale if clients respect the rules
+- REST focuses on uniformity: if you’re using HTTP you should utilize HTTP features whenever possible, instead of inventing conventions
+
+These constraints of REST when applied to HTTP APIs can help the API last for decades, which is a whole lot more complex without these concepts.
+
+REST does not require the use of schema metadata (like WSDL or similar), but does allow it. In fact REST has no opinions either way: it does not explicitly demand it, nor disallow it. the metadata is something many API developers hated about SOAP, and these days it is once again all the rage in gRPC and GraphQL, but REST has a few type systems available for optional use, the main one these days being: JSON Schema.
+
+JSON Schema is inspired by XML Schema — but not functionally identical — and is one of the most important things to happen to HTTP APIs in years, and will be discussed a lot throughout the book.
+
+Unfortunately, REST become a marketing buzzword for most of 2006–2014. It became a metric of quality that developers would aspire to, fail to understand, then label as REST anyway. Most systems saying they are REST are little more than RPC with HTTP verbs and pretty URLs. As such, you might not get cacheability provided, it might have a bunch of wacky conventions, and there might not be any links for you to use to discover next available actions. These APIs are jokingly called REST__ish__ by people aware of the difference.
```

Comment:
```suggestion
Unfortunately, REST became a marketing buzzword for most of 2006–2014. It became a metric of quality that developers would aspire to, fail to understand, then label as REST anyway. Most systems saying they are REST are little more than RPC with HTTP verbs and pretty URLs. As such, you might not get cacheability provided, it might have a bunch of wacky conventions, and there might not be any links for you to use to discover next available actions. These APIs are jokingly called REST__ish__ by people aware of the difference.
```